File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/Mercado.java
package com.mobnetic.coinguardian.model.market;

import java.util.HashMap;
import java.util.LinkedHashMap;

import org.json.JSONObject;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardian.model.currency.Currency;
import com.mobnetic.coinguardian.model.currency.VirtualCurrency;
import com.mobnetic.coinguardian.util.TimeUtils;

public class Mercado extends Market {

	private final static String NAME = "Mercado Bitcoin";
	private final static String TTS_NAME = "Mercado";
	private final static String URL_BTC = "https://www.mercadobitcoin.com.br/api/ticker/";
	private final static String URL_LTC = "https://www.mercadobitcoin.com.br/api/ticker_litecoin/";
	private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<String, CharSequence[]>();
	static {
		CURRENCY_PAIRS.put(VirtualCurrency.BTC, new String[]{
				Currency.BRL
			});
		CURRENCY_PAIRS.put(VirtualCurrency.LTC, new String[]{
				Currency.BRL
			});
	}
	
	public Mercado() {
		super(NAME, TTS_NAME, CURRENCY_PAIRS);
	}

	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return VirtualCurrency.LTC.equals(checkerInfo.getCurrencyBase()) ? URL_LTC : URL_BTC;
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		final JSONObject tickerJsonObject = jsonObject.getJSONObject("ticker");
		
		ticker.bid = tickerJsonObject.getDouble("buy");
		ticker.ask = tickerJsonObject.getDouble("sell");
		ticker.vol = tickerJsonObject.getDouble("vol");
		ticker.high = tickerJsonObject.getDouble("high");
		ticker.low = tickerJsonObject.getDouble("low");
		ticker.last = tickerJsonObject.getDouble("last");
		ticker.timestamp = tickerJsonObject.getLong("date")*TimeUtils.MILLIS_IN_SECOND;
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Mercado.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.model.currency.Currency
import com.aneonex.bitcoinchecker.datamodule.model.currency.CurrencyPairsMap
import com.aneonex.bitcoinchecker.datamodule.model.currency.VirtualCurrency
import com.aneonex.bitcoinchecker.datamodule.util.TimeUtils
import org.json.JSONObject

class Mercado : Market(NAME, TTS_NAME, getCurrencyPairs()) {
    companion object {
        private const val NAME = "Mercado Bitcoin"
        private const val TTS_NAME = "Mercado"
        private const val URL = "https://www.mercadobitcoin.net/api/%1\$s/ticker/"

        @Suppress("SpellCheckingInspection")
        private fun getCurrencyPairs(): CurrencyPairsMap {
            // API Doc: https://www.mercadobitcoin.com.br/api-doc/
            val baseCurrencies = arrayOf(
                    "ASRFT",
                    "ATMFT",
                    VirtualCurrency.BCH,
                    VirtualCurrency.BTC,
                    "CAIFT",
                    "CHZ",
                    VirtualCurrency.ETH,
                    "GALFT",
                    "IMOB01",
                    "JUVFT",
                    VirtualCurrency.LINK,
                    VirtualCurrency.LTC,
                    "MBCONS01",
                    "MBCONS02",
                    "MBFP01",
                    "MBVASCO01",

                    "MBPRK01",
                    "MBPRK02",
                    "MBPRK03",
                    "MBPRK04",

                    "PAXG",
                    "PSGFT",
                    VirtualCurrency.USDC,
                    "WBX",
                    VirtualCurrency.XRP,
            )

            val quoteCurrencies = arrayOf(Currency.BRL)
            return baseCurrencies.associateTo(CurrencyPairsMap()) { it to quoteCurrencies }
        }
    }

    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyBase)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        val tickerJsonObject = jsonObject.getJSONObject("ticker")
        ticker.bid = tickerJsonObject.getDouble("buy")
        ticker.ask = tickerJsonObject.getDouble("sell")
        ticker.vol = tickerJsonObject.getDouble("vol")
        ticker.high = tickerJsonObject.getDouble("high")
        ticker.low = tickerJsonObject.getDouble("low")
        ticker.last = tickerJsonObject.getDouble("last")
        ticker.timestamp = tickerJsonObject.getLong("date") * TimeUtils.MILLIS_IN_SECOND
    }
}